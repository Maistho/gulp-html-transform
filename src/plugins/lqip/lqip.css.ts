export const styles = `
.lqip {
  position: relative;
  width: 100%;
  overflow: hidden;
  background-size: 0 0;
  background-repeat: no-repeat;
  display: inline-block;
  transition: opacity .5s ease-in-out;
}

.lqip::after {
  content: "";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  opacity: 0;
  transition: inherit;
  z-index: 1;
  background-image: inherit;
  background-size: cover;
  background-repeat: no-repeat;
  filter: blur(15px);
  transform: scale(1.1);
}

.lqip.blur::after {
  opacity: 1;
}

.lqip img {
  transition: inherit;
  display: block;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}
`
