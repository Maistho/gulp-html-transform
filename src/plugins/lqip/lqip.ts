const chalk = require('chalk')
import * as path from 'path'
import * as fs from 'fs'
import { promisify } from 'util'

import { Transformer } from '../../transform'
import { styles } from './lqip.css'

export interface LqipOptions {
  /**
   * @deprecated not needed since 2.1.1, paths are calculated based on the file paths
   */
  base?: string

  /**
   * custom query for finding images
   * @default img[src]
   */
  query?: string

  /**
   * Whether to add the styles to the <head> of your document
   */
  addStyles?: boolean
}

export const lqip = (options: LqipOptions): Transformer => {
  let base64: any
  let sizeOf: any
  try {
    base64 = require('lqip').base64
    sizeOf = promisify(require('image-size'))
  } catch (err) {
    console.warn(`lqip requires ${chalk.red('lqip')} and ${chalk.red('image-size')} to be installed`)
    return async () => {
      // noop
    }
  }

  options = Object.assign({
    query: 'img[src]',
  }, options)

  return async ($: CheerioStatic, { dirname }) => {
    const promises: Promise<void>[] = []

    if (options.addStyles) {
      $('head').prepend(`<style>${styles}</style>`)
    }

    const elements = $(options.query).toArray().map(el => $(el))

    for (const $el of elements) {
      const src = $el.attr('src')
      const re = /^https?:\/\//i
      if (re.test(src)) {
        return
      }
      const filepath = path.join(options.base || dirname, src)

      const p = Promise.all([
        base64(filepath),
        sizeOf(filepath),
      ]).then(([res, dimensions]) => {
        const wrapper = $('<div />')
        wrapper.css('padding-top', ((dimensions.height / dimensions.width) * 100).toFixed(4) + '%')
        wrapper.css('background-image', `url(${res})`)
        wrapper.attr('class', `lqip blur ${$el.attr('class')}`)
        const clone = $el.clone()
        clone.attr('onload', 'this.parentElement.classList.remove(\'blur\')')
        clone.attr('class', '')
        wrapper.append(clone)
        $el.replaceWith(wrapper)
      }).catch(err => {
        console.warn(`lqip: Received an error when trying to load ${chalk.red(filepath)}`, err)
        throw err
      })
      promises.push(p)
    }

    await Promise.all(promises)
  }
}
