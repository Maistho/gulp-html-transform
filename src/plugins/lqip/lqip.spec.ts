import { lqip } from './lqip'

import { testTransform } from '../../test/helpers'

describe('lqip', () => {
  const filename = 'image.html'
  it('creates a base64 lqip', async () => {
    const content = await testTransform(filename, [lqip()])
    expect(content).toMatchSnapshot()
  })
  it('creates a background-color', async () => {
    const content = await testTransform(filename, [
      lqip({ method: 'primaryColor' }),
    ])
    expect(content).toMatchSnapshot()
  })
  it('adds styles to the head', async () => {
    const content = await testTransform(filename, [
      lqip({ query: '.noop', addStyles: true }),
    ])
    expect(content).toMatchSnapshot()
  })
})
