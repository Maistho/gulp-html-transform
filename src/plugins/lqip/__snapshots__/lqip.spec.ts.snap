// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lqip adds styles to the head 1`] = `
"<html><head><style>
.lqip {
  position: relative;
  width: 100%;
  overflow: hidden;
  background-size: 0 0;
  background-repeat: no-repeat;
  display: inline-block;
  transition: opacity .5s ease-in-out;
}

.lqip::after {
  content: \\"\\";
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  opacity: 0;
  transition: inherit;
  z-index: 1;
  background-color: inherit;
  background-image: inherit;
  background-size: cover;
  background-repeat: no-repeat;
  filter: blur(15px);
  transform: scale(1.1);
}

.lqip.blur::after {
  opacity: 1;
}

.lqip img {
  transition: inherit;
  display: block;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}
</style></head>
  <body>
    <img src=\\"test.jpg\\">
  

</body></html>"
`;

exports[`lqip creates a background-color 1`] = `
"<html><head></head>
  <body>
    <div style=\\"padding-top: 66.6667%; background-color: #063559;\\" class=\\"lqip blur\\"><img src=\\"test.jpg\\" onload=\\"this.parentElement.classList.remove(&apos;blur&apos;)\\"></div>
  

</body></html>"
`;

exports[`lqip creates a base64 lqip 1`] = `
"<html><head></head>
  <body>
    <div style=\\"padding-top: 66.6667%; background-image: url(data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAf/AABEIAAcACgMBEQACEQEDEQH/xAGiAAABBQEBAQEBAQAAAAAAAAAAAQIDBAUGBwgJCgsQAAIBAwMCBAMFBQQEAAABfQECAwAEEQUSITFBBhNRYQcicRQygZGhCCNCscEVUtHwJDNicoIJChYXGBkaJSYnKCkqNDU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6g4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2drh4uPk5ebn6Onq8fLz9PX29/j5+gEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoLEQACAQIEBAMEBwUEBAABAncAAQIDEQQFITEGEkFRB2FxEyIygQgUQpGhscEJIzNS8BVictEKFiQ04SXxFxgZGiYnKCkqNTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqCg4SFhoeIiYqSk5SVlpeYmZqio6Slpqeoqaqys7S1tre4ubrCw8TFxsfIycrS09TV1tfY2dri4+Tl5ufo6ery8/T19vf4+fr/2gAMAwEAAhEDEQA/APNfGn/BT7XPjXbXGs6H8IvgTc6jqG1U1HVvhHBZorXTqscl0uleJtHuHtIJZF3RJHO8oijdrY5dBrwHx3xJwbVpwx3HHGuKy9TlUhhcJj8qwsaUKMpKXJh4ZD7FttcseWrSlG/u1EtTr444R4f46hUvwJwFPNfYwoPNs4yfMcxxU/b8riqmK/t6jX9nFS5pKVOsp681GUtF8TXf7QPxkubu6ubb4VfCrULa4uJp7e/tPB9hZWl9BLI0kV5a2d18S7a5tLe5jZZoba4t4J4I3WKWGKRGRf6tpfSMw8KVKE80z6MoU4RlGeYYuUoyjFJxlKOXOMpJppyi3Fu7TaP5mq/RfU6tSayXhiSnUnJSo5Jl0aUuaTfNSjUzKFSNN3vCM4Qmo2UoxldL/9k=);\\" class=\\"lqip blur\\"><img src=\\"test.jpg\\" onload=\\"this.parentElement.classList.remove(&apos;blur&apos;)\\"></div>
  

</body></html>"
`;
